publisher/release/marker_aws.go

```diff
@@ -0,0 +1,200 @@
+var ErrNoStartedMarkersFound = errors.New("no started markers found")
+var ErrNoStartedMarkerFoundForApp = errors.New("no started marker found for app")
+var ErrCannotWriteMarkerFile = errors.New("cannot write marker file")
+var ErrNotStartedMark = errors.New("not started mark")
+
+// S3Config markerAWS lock config DTO.
+type S3Config struct {
+	Directory string
+	Bucket    string
+	RoleARN   string
+	Region    string
+}
+
+// S3Client aws client interface for testing
+type S3Client interface {
+	PutObject(input *s3.PutObjectInput) (*s3.PutObjectOutput, error)
+	GetObject(input *s3.GetObjectInput) (*s3.GetObjectOutput, error)
+}
+
+type TimeProvider interface {
+	Now() time.Time
+}
+
+type RealTimeProvider struct{}
+
+func (RealTimeProvider) Now() time.Time {
+	return time.Now()
+}
+
+type markerAWS struct {
+	client       S3Client
+	conf         S3Config
+	timeProvider TimeProvider
+	logfn        Logf
+}
+
+// NewMarkerAWS creates a new marker using AWS S3
+// it returns an interface on purpose, so this way
+// we can have markerAWS unexported and force the
+// usage of the constructor
+func NewMarkerAWS(s3Config S3Config, logfn Logf) (Marker, error) {
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+
+	creds := stscreds.NewCredentials(sess, s3Config.RoleARN, func(p *stscreds.AssumeRoleProvider) {})
+	awsCfg := aws.Config{
+		Credentials: creds,
+		Region:      aws.String(s3Config.Region),
+	}
+
+	return &markerAWS{
+		client:       s3.New(sess, &awsCfg),
+		conf:         s3Config,
+		timeProvider: RealTimeProvider{},
+		logfn:        logfn,
+	}, nil
+}
+
+// Start will:
+// load all the markers from the file
+// append a new started marker
+// write the markers back to the file
+func (s *markerAWS) Start(releaseInfo ReleaseInfo) (Mark, error) {
+	s.logfn("[marker] starting %s", releaseInfo.AppName)
+	markers, err := s.readMarkers()
+	if err != nil {
+		// Do not return error if the file does not exist. The first time we will create a new one.
+		if !isNoSuchKeyError(err) {
+			return Mark{}, err
+		}
+	}
+
+	mark := Mark{
+		ReleaseInfo: releaseInfo,
+		Start:       CustomTime{s.now()},
+	}
+
+	markers = append(markers, mark)
+	if err = s.writeMarkers(markers); err != nil {
+		return mark, err
+	}
+
+	return mark, nil
+}
+
+// End will:
+// load all the markers from the file
+// find the last started marker
+// append the end time to the last started marker
+// write the markers back to the file
+func (s *markerAWS) End(mark Mark) error {
+	s.logfn("[marker] ending %s", mark.AppName)
+	if mark.Start.IsZero() {
+		return ErrNotStartedMark
+	}
+
+	markers, err := s.readMarkers()
+	if err != nil {
+		return err
+	}
+	//ensure the latest marker is the one being ended
+	if len(markers) == 0 {
+		return ErrNoStartedMarkersFound
+	}
+
+	lastMarker := markers[len(markers)-1]
+	if !lastMarker.End.IsZero() {
+		return ErrLastMarkerEnded
+	}
+
+	if lastMarker.AppName != mark.AppName || !lastMarker.Start.Equals(mark.Start) {
```

Comment:
Are packages uploaded 1 by 1 always? What if multiple teams try to release at the same time? This tool doesn't allow for it. Should this be allowed in the future?

Reply:
There is a locking mechanism to prevent multiple actions running at the same time within the same S3 Bucket.

https://github.com/newrelic/infrastructure-publish-action/blob/main/publisher/lock/lock.go